src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java

```diff
@@ -94,16 +141,38 @@
                 refEx = answer.refEx;
             }
 
+            this.enumCtx = new EnumCtx(homeCtx, answer, homeCtx.clnt);
             // Ensures that context won't get closed from underneath us
-            this.homeCtx = homeCtx;
-            homeCtx.incEnumCount();
-            enumClnt = homeCtx.clnt; // remember
+            this.enumCtx.homeCtx.incEnumCount();
+            this.cleanable = CleanerFactory.cleaner().register(this, this.enumCtx);
+            // For finalizers, JLS guarantees that the constructor of a finalizeable
+            // class will complete before the finalizer is run. Classes that use
+            // Cleaner instead don't currently have such a guarantee.
+
+            // 'this' needs to stay reachable until registration with Cleaner
+            // completes, to ensure that the cleanable state that has been created
+            // will eventually be cleaned up. Within Cleaner.register(), there
+            // are reachability fences to ensure that the registered object remains
+            // reachable.
+            // TODO: Is anything else needed so that this constructor
+            //  "happens-before" the cleaning action ?
```

Comment:
I think the resolution of this TODO is to modify the specification of Cleaner to require that:

1. the object being registered is guaranteed reachable until after registration has completed; and
2. a "memory consistency effects" clause should also be added to Cleaner.register. For an example, see the class specification of CopyOnWriteArrayList.

It may be that reachabilityFences are already in the right place in Cleaner. Great. But last time I looked it doesn't have any memory visibility operations. I'd suggest opening a bug on Cleaner for the spec change and accumulating notes there.